src/main.cpp

```diff
@@ -18,104 +18,183 @@
+	for (int i = 0; i < argc; i++) args.emplace_back(argv[i]);
+	return args;
 }
 
-//
-// Playd
-//
-
-int Playd::GetDeviceID()
+/**
+ * Tries to get the output device ID from program arguments.
+ * @param args The program argument vector.
+ * @return The device ID, -1 if invalid selection (or none).
+ */
+int GetDeviceID(const std::vector<std::string> &args)
 {
 	// Did the user provide an ID at all?
-	if (this->argv.size() < 2) return this->INVALID_ID;
+	if (args.size() < 2) return -1;
 
 	// Only accept valid numbers (stoi will throw for invalid ones).
 	int id;
 	try {
-		id = std::stoi(this->argv.at(1));
+		id = std::stoi(args.at(1));
 	} catch (...) {
 		// Only std::{invalid_argument,out_of_range} are thrown here.
-		return this->INVALID_ID;
+		return -1;
 	}
 
 	// Only allow valid, outputtable devices; reject input-only devices.
-	if (!this->audio.IsOutputDevice(id)) return this->INVALID_ID;
+	if (!SdlAudioSink::IsOutputDevice(id)) return -1;
 
 	return id;
 }
 
-Playd::Playd(int argc, char *argv[])
-    : audio(),
-      pfile(this, audio),
-      pposition(this, Playd::POSITION_PERIOD),
-      pstate(this),
-      player(this, pfile, pposition, pstate),
-      handler(player)
-{
-	for (int i = 0; i < argc; i++) {
-		this->argv.emplace_back(argv[i]);
-	}
+/**
+ * Sets up the audio system with the desired sources and sinks.
+ * @param audio The audio system to configure.
+ */
+void SetupAudioSystem(PipeAudioSystem &audio) {
+	audio.SetSink(&SdlAudioSink::Build);
+
+// Now set up the available sources.
+#ifdef WITH_FLAC
+	audio.AddSource("flac", &FlacAudioSource::Build);
+#endif // WITH_FLAC
+
+#ifdef WITH_MP3
+	mpg123_init();
+	atexit(mpg123_exit);
+	audio.AddSource("mp3", &Mp3AudioSource::Build);
+#endif // WITH_MP3
+
+#ifdef WITH_SNDFILE
+	audio.AddSource("flac", &SndfileAudioSource::Build);
+	audio.AddSource("ogg", &SndfileAudioSource::Build);
+	audio.AddSource("wav", &SndfileAudioSource::Build);
+#endif // WITH_SNDFILE
 }
 
-int Playd::Run()
-{
-	// Fill in some default arguments.
-	// Note that we don't have a default device ID; if the user doesn't
-	// supply an ID, we treat it as if they had supplied an invalid one.
-	auto size = this->argv.size();
-	std::string addr = size > 2 ? this->argv.at(2) : "0.0.0.0";
-	std::string port = size > 3 ? this->argv.at(3) : "1350";
-
-	// Now set up the device ID.
-	// Do this now, so that an invalid ID is caught before we start trying
-	// to acquire the network socket.
-	int id = this->GetDeviceID();
-	if (id == INVALID_ID) {
-		// Show the user the valid device IDs they can use.
-		auto device_list = this->audio.GetDevicesInfo();
-		for (const auto &device : device_list) {
-			std::cout << device.first << ": " << device.second
-			          << std::endl;
-		}
-		return EXIT_FAILURE;
+/**
+ * Reports usage information and exits.
+ * @param progname The name of the program as executed.
+ */
+ [[noreturn]] void ExitWithUsage(const std::string &progname) {
```

Comment:
I assumed it was more useful for documentation purposes here, just to say 'if you hadn't realised, this doesn't return'